### PR TITLE
Fix RobustChannel not reconnecting on duplicate named queue/exchanges

### LINKED ISSUE #229: RobustChannel will not reconnect if the queue/exchange is declared twice with the same name.
If you declare exchange for the second time (with the same parameters), it overrides the previous exchange in `self._exchanges[name]` reference, which causes "on_reconnect" not to pass new channel to declared exchanges/queues, resulting in `RuntimeError: <Channel: "1"> closed` errors.

I would make a pull request but don't know what's the correct solution, so I will propose few below:

I will state the examples on **exchanges**, but everything also applies to **queues**

**Option 1 - Fetch exchange from already declared dict**

`robust_channel.py`
```
async def declare_exchange(self, name: str,
                           type: ExchangeType=ExchangeType.DIRECT,
                           durable: bool=None, auto_delete: bool=False,
                           internal: bool=False, passive: bool=False,
                           arguments: dict=None, timeout: int=None,
                           robust: bool=True) -> Exchange:

+   if name in self._exchanges:
+       return self._exchanges[name]

    exchange = await super().declare_exchange(
        name=name, type=type, durable=durable, auto_delete=auto_delete,
        internal=internal, passive=passive, arguments=arguments,
        timeout=timeout,
    )

    if not internal and robust:
        self._exchanges[name] = exchange

    return exchange
```

Cons: 
- You return object which is used elsewhere
- If you declare exchange with different parameters it will silently pass and return current instance, instead of raising exception.

**Option 2 - variance on Option 1**
Actually trying to declare the exchange and discarting fetched object afterward 
`robust_channel.py`
```
async def declare_exchange(self, name: str,
                           type: ExchangeType=ExchangeType.DIRECT,
                           durable: bool=None, auto_delete: bool=False,
                           internal: bool=False, passive: bool=False,
                           arguments: dict=None, timeout: int=None,
                           robust: bool=True) -> Exchange:

    exchange = await super().declare_exchange(
        name=name, type=type, durable=durable, auto_delete=auto_delete,
        internal=internal, passive=passive, arguments=arguments,
        timeout=timeout,
    )

+   if name in self._exchanges:
+       return self._exchanges[name]

    if not internal and robust:
        self._exchanges[name] = exchange

    return exchange
```

Cons: 
- You return object which is used elsewhere
- Discarting awaited object.
- Confusing code without explanation/comments

**Option 3 - Compare all exchange connection parameters**
`robust_channel.py`
```
async def declare_exchange(self, name: str,
                           type: ExchangeType=ExchangeType.DIRECT,
                           durable: bool=None, auto_delete: bool=False,
                           internal: bool=False, passive: bool=False,
                           arguments: dict=None, timeout: int=None,
                           robust: bool=True) -> Exchange:

+   if name in self._exchanges:
+       exchange = self._exchanges[name]
+       if exchange.__type != type or exchange.durable is not durable or ... # rest
+           raise RuntimeException("can't declare exchange... different arguments") 
+       return self._exchanges[name]

    exchange = await super().declare_exchange(
        name=name, type=type, durable=durable, auto_delete=auto_delete,
        internal=internal, passive=passive, arguments=arguments,
        timeout=timeout,
    )

    if not internal and robust:
        self._exchanges[name] = exchange

    return exchange
```

**Option 4 - Store all the created exchanges**

`robust_channel.py`
```
async def declare_exchange(self, name: str,
                           type: ExchangeType=ExchangeType.DIRECT,
                           durable: bool=None, auto_delete: bool=False,
                           internal: bool=False, passive: bool=False,
                           arguments: dict=None, timeout: int=None,
                           robust: bool=True) -> Exchange:

    exchange = await super().declare_exchange(
        name=name, type=type, durable=durable, auto_delete=auto_delete,
        internal=internal, passive=passive, arguments=arguments,
        timeout=timeout,
    )

    if not internal and robust:
+       self._exchanges[(name, id(exchange))] = exchange

    return exchange
```

Cons:
- `exchange_delete` can't be done just by name - since you can have multiple instance with same name 
- OR delete all the references to the exchange with the name, in which case either make self._exchanges[name] = list() for easier delete lookup, but harder to cycle through (2 levels of nesting)

I'm inclined to Option 2 or Option 3, but I'd like to know other opinions or Options.

### PR DESCRIPTION
**Pull request by aviramha**
Fixes #229 
I wasn't sure if to add tests as my approach is pretty simple...

### CODE REVIEW

File: aio_pika/robust_channel.py [hunk 1]
```diff
@@ -42,8 +43,8 @@ def __init__(self, connection, channel_number: int = None,
             on_return_raises=on_return_raises,
         )
 
-        self._exchanges = dict()
-        self._queues = dict()
+        self._exchanges = []
```
Comment (mosquito) [inline aio_pika/robust_channel.py:46]: `defaultdict(set)` will looks better I guess.

Comment (aviramha) [reply aio_pika/robust_channel.py:46]: But why should it be a dict?

File: aio_pika/robust_channel.py [hunk 2]
```diff
@@ -121,7 +122,8 @@ def _on_channel_close(self, exc: Exception):
             if_unused=if_unused, nowait=nowait
         )
 
-        self._exchanges.pop(exchange_name, None)
+        self._exchanges[:] = filterfalse(lambda x: x.name == exchange_name,
```
Comment (mosquito) [inline aio_pika/robust_channel.py:125]: This change unneeded when `self._exchanges` is an instance of dict (defaultdict)

Comment (aviramha) [reply aio_pika/robust_channel.py]: I think it's less nice then just a list, but OK. updated the PR

File: aio_pika/robust_channel.py [hunk 3]
```diff
@@ -108,7 +111,7 @@ def _on_channel_close(self, exc: Exception):
         )
 
         if not internal and robust:
-            self._exchanges[name] = exchange
+            self._exchanges.append(exchange)
```
Comment (mosquito) [inline aio_pika/robust_channel.py:114]: `self._exchanges[name].add(exchange)`

File: aio_pika/robust_channel.py [hunk 4]
```diff
@@ -121,7 +124,7 @@ def _on_channel_close(self, exc: Exception):
             if_unused=if_unused, nowait=nowait
         )
 
-        self._exchanges.pop(exchange_name, None)
+        self._exchanges.pop(exchange_name)
```
Comment (mosquito) [inline aio_pika/robust_channel.py:127]: revert it please

File: aio_pika/robust_channel.py [hunk 5]
```diff
@@ -150,7 +153,8 @@ def _on_channel_close(self, exc: Exception):
             if_unused=if_unused, if_empty=if_empty, nowait=nowait
         )
 
-        self._queues.pop(queue_name, None)
+        self._queues.pop(queue_name)
```
Comment (mosquito) [inline aio_pika/robust_channel.py:156]: revert it please

**Comment by aviramha** [comment]
Done

**Comment by mosquito** [comment]
Looks simple and nice, @decaz @dizballanze, please review.

**Comment by dizballanze** [comment]
Hi, @aviramha. Maybe you could write a test for this?

**Comment by aviramha** [comment]
Master is broken or something? The tests fail on chunks of code that existed beforehand.

**Comment by dizballanze** [comment]
@mosquito

**Comment by aviramha** [comment]
@mosquito Any ETA on the next version release? :)

**Comment by mosquito** [comment]
[6.6.1](https://pypi.org/project/aio-pika/6.6.1/) has been released

**Comment by aviramha** [comment]
For some reason I can't find on master the changes I've done and it's not in the changel log also. Was the PR reverted?

**Comment by mosquito** [comment]
See #323 I'm applied your original commit so you are co-author

**Comment by aviramha** [comment]
Ah I thought the fix is in master already.

**Comment by Sinkler** [comment]
Hello!

After this changing we detected a memory leak in our project. It looks like there is a problem in `defaultdict(set)`. We replaced it by `defaultdict(WeakSet)` and now it's ok:

```python
class FixRobustChannel(RobustChannel):
    def __init__(
        self,
        connection,
        channel_number: int = None,
        publisher_confirms: bool = True,
        on_return_raises=False,
    ):
        super().__init__(connection, channel_number, publisher_confirms, on_return_raises)
        self._exchanges = defaultdict(WeakSet)
        self._queues = defaultdict(WeakSet)


class FixRobustConnection(RobustConnection):
    CHANNEL_CLASS = FixRobustChannel


await aio_pika.connect_robust("...", connection_class=FixRobustConnection)
```

**Pull request merged**